### PR TITLE
Fix the handling for concurrent queries over UDP

### PR DESCRIPTION
**Pull request by sanimej**
For queries forwarded over a cached connection to external servers maintain the context for the queries so that reply from the external server can be sent back to the right client connection.

Signed-off-by: Santhosh Manohar <santhosh@docker.com>

### CODE REVIEW

File: resolver.go [hunk 1]
```diff
@@ -418,22 +436,69 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
...
 	}
-	r.count--
-	return true
+
+	switch proto {
+	case "tcp":
+		break
+	case "udp":
+		if cc, ok = r.client[msg.Id]; ok == false {
+			w = nil
```
Comment (mrjana) [inline resolver.go]: You should just `return nil` here. There is no point in proceeding since the entry is not in the map.

File: resolver.go [hunk 2]
```diff
@@ -418,22 +436,69 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
...
+		}
+		delete(r.client, msg.Id)
+		msg.Id = cc.dnsID
+		w = cc.respWriter
+	default:
+		w = nil
```
Comment (mrjana) [inline resolver.go]: Again just return here with nil

**Comment by sanimej** [comment]
@mrjana Thanks. Addressed the comment. PTAL.

**Comment by mrjana** [comment]
LGTM

**Comment by aboch** [comment]
LGTM

**Pull request merged**